test/test_sparse.py

```diff
@@ -2881,6 +2882,136 @@ def sparse_log(x):
+                result = mat1.dot(mat2).tocoo()
+                return torch.sparse_coo_tensor([result.row, result.col], result.data, result.shape)
+
+            except ImportError as msg:
+                assert a.shape[1] == b.shape[0]
+                n, p = a.shape
+                p, m = b.shape
+                indices_a = a._indices()
+                values_a = a._values()
+                indices_b = b._indices()
+                values_b = b._values()
+                nnz1 = len(indices_a[0])
+                nnz2 = len(indices_b[0])
+
+                if a.is_coalesced() and b.is_coalesced():
+                    r2 = _indices2csr(indices_b[0], b.shape[0])
+                    d = defaultdict(values_b.numpy().dtype.type)
+                    for n1 in range(nnz1):
+                        for n2 in range(r2[indices_a[1][n1]], r2[indices_a[1][n1] + 1]):
+                            d[indices_a[0][n1].item(), indices_b[1][n2].item()] += values_a[n1] * values_b[n2]
+
+                else:
+                    d = defaultdict(values_b.numpy().dtype.type)
+                    for n1 in range(nnz1):
+                        for n2 in range(nnz2):
+                            if indices_b[0][n2] == indices_a[1][n1]:
+                                d[indices_a[0][n1].item(), indices_b[1][n2].item()] += values_a[n1] * values_b[n2]
+                i3 = []
+                j3 = []
+                values = []
+                for i, j in sorted(d):
+                    i3.append(i)
+                    j3.append(j)
+                    values.append(d[i, j])
+                return torch.sparse_coo_tensor(torch.tensor([i3, j3]), torch.tensor(values), (n, m))
+
+        def test_sparse_matmul_grad(sparse_dims, nnz, shape_a, shape_b):
+            a, i_a, v_a = self._gen_sparse(sparse_dims, nnz, shape_a)
+            b, i_b, v_b = self._gen_sparse(sparse_dims, nnz, shape_b)
+            a.requires_grad = True
+            b.requires_grad = True
+            a_dense = a.to_dense().detach()
+            b_dense = b.to_dense().detach()
+            a_dense.requires_grad = True
+            b_dense.requires_grad = True
+            r1 = a_dense.matmul(b_dense)
+            r2 = torch.sparse.mm(a, b)
+            f1 = torch.sum(r1)
+            f1.backward()
+
+            f2 = torch.sparse.sum(r2)
+            f2.backward()
+            self.assertEqual(a.grad.to_dense(), a_dense.grad)
+            self.assertEqual(b.grad.to_dense(), b_dense.grad)
+
+        def test_sparse_matmul(sparse_dims, nnz, shape_a, shape_b):
+            a, i_a, v_a = self._gen_sparse(sparse_dims, nnz, shape_a)
+            b, i_b, v_b = self._gen_sparse(sparse_dims, nnz, shape_b)
+
+            # python implementation
+            r1 = sparse_mm(a, b)
+
+            self.assertEqual(r1.to_dense(), torch.mm(a.to_dense(), b.to_dense()))
+            self.assertEqual((r1._values() != 0).all(), True)
```

Comment:
There exists a chance that `_gen_sparse` will generate `a`, `b` such that this assert will fail. In general, it is ok to have zeros in `values`. I suggest removing this test.